data/tutorials/lg_08_error_handling.md

```diff
@@ -9,62 +9,127 @@ date: 2021-05-27T21:07:30-00:00
 
 # Error Handling
 
+## Error as Special Values
+
+Don't do that.
+
+Some languages, most emblematically C, treats certain values as errors. For
+instance, when receiving data through a network connection, a function expected
+to return the number of received bytes might return a negative number meaning:
+“timed out waiting”. Another example would be returning the empty string when
+extracting a substring of negative length. Great software was written using this
+style, but is is not the proper way to deal with errors in OCaml.
+
+OCaml has three major ways to deal with errors:
+1. Exceptions
+1. `Option` values
+1. `Result` values
+
+Use them. Do not encode errors inside data. 
+
+Exceptions provide a mean to deal
+with errors at the control flow level while `Option` and `Result` provide a mean
+to turn errors into dedicated data.
+
+The rest of this document presents and compares approaches towards error
+handling.
+
 ## Exceptions
 
-One way of handling errors in OCaml is exceptions. The
+Historically, the first way of handling errors in OCaml is exceptions. The
 standard library relies heavily upon them.
 
-Exceptions belong to the type `exn` (an extensible sum type):
+The biggest issue with exceptions is that they do not appear in types. One has
+to read the documentation to see that, indeed, `List.find` or `String.sub` are
+not total functions, and that they might fail by raising an exception.
+
+However, exceptions have the great merit of being compiled into efficient
+machine code. When implementing trial and error approaches likely to back-track
+often, exceptions can be used to acheive good performance.
+
+Exceptions belong to the type `exn` which is an [extensible sum
+type](/releases/latest/manual/extensiblevariants.html).
 
 ```ocaml
+# exception Foo of string;;
 exception Foo of string
 
-let i_will_fail () =
-  raise (Foo "Oh no!")
+# let i_will_fail () =
+  raise (Foo "Oh no!");;
+
+# i_will_fail ();;
```

Comment:
The output I get here is `val i_will_fail : unit -> 'a = <fun>`
Without the Exception on li 61